manifests/npm.pp

```diff
@@ -1,58 +1,88 @@
-# Define: nodejs::npm
-#
-# Parameters:
-#
-# Actions:
-#
-# Requires:
-#
-# Usage:
-#
+# See README.md for usage information.
 define nodejs::npm (
-  $ensure      = present,
-  $version     = undef,
-  $source      = undef,
-  $install_opt = undef,
-  $remove_opt  = undef
+  $ensure            = 'present',
+  $install_options   = [],
+  $npm_path          = $nodejs::params::npm_path,
+  $npm_tag           = undef,
+  $source            = undef,
+  $uninstall_options = [],
 ) {
-  include nodejs
 
-  $npm = split($name, ':')
-  $npm_dir = $npm[0]
-  $npm_pkg = $npm[1]
+  $npm_array = split($name, '[|]')
 
+  $npm_package = $npm_array[0]
+  $install_dir = $npm_array[1]
+
+  validate_re($ensure, '^[^<>=]', "The ${module_name}::npm defined type does not accept version ranges")
+  validate_absolute_path($install_dir)
+  validate_array($install_options)
+  validate_string($npm_package)
+  validate_array($uninstall_options)
+
+  $install_options_string = join($install_options, ' ')
+  $uninstall_options_string = join($uninstall_options, ' ')
+
+  # install_check will always return false when source is provided
   if $source {
-    $install_pkg = $source
-  } elsif $version {
-    $install_pkg = "${npm_pkg}@${version}"
-  } else {
-    $install_pkg = $npm_pkg
+    $package_string = $source
+  }
+  # install_check will always return false when tag is provided
+  # npm ls does not keep track of tags after install
+  elsif $npm_tag {
+    $package_string = "${npm_package}@${npm_tag}"
+  }
+  elsif $ensure == 'latest' {
+    $package_string = "${npm_package}@latest"
```

Comment:
if that was the case, we could, again, collapse these two branches, i.e.: if ensure matches latest, or a version string, just append it.
in that case, it would make sense to make this the default, last case. i.e.: present, absent, etc.. have already been handled, now we just give up and decide it's a version number (or latest)